src/rulesets/LinearAlgebra/utils.jl

```diff
@@ -43,3 +43,36 @@ Symmetric
 ````
 """
 _unionall_wrapper(::Type{T}) where {T} = T.name.wrapper
+
+"""
+    WithSomeZeros{T}
+
+This is a union of LinearAlgebra types, all of which are partly structral zeros,
+with a simple backing array given by `parent(x)`. All have methods of `_rewrap`
+to re-create.
+
+This exists to solve a type instability, as broadcasting for instance
+`λ .* Diagonal(rand(3))` gives a dense matrix when `x==Inf`.
+But `withsomezeros_rewrap(x, λ .* parent(x))` is type-stable.
+"""
+WithSomeZeros{T} = Union{
```

Comment:
I would call these StructuredSparseArray

Reply:
You approve of the mechanism, https://github.com/JuliaDiff/ChainRules.jl/pull/337#issuecomment-754252513?

Reply:
I am willing to give it a shot.
We can always change it later.
It's not going to lead to wrong behavour AFAICT.

It seems unfortunate not to take advantage of the fact that we know where the zeros are,
and we know that the pullback is going to map zeros to zeros, since linear.
So we should be able to skip some.
But idk that that is a generic API for our structurally sparse matrixes to know if an index will be zero.

Reply:
Maybe I misunderstand you, but both `λ .* Diagonal(rand(3))` and this function do know where the zeros are, and do `O(N)` work. That's the only really sparse one. 

For UpperTriangular, I haven't tried to time this against broadcasting... there could be trade-offs. Frankly I doubt that anyone has ever called `norm(::UpperTriangular)` outside a test, though. So perhaps thinking about that can wait until this finds wider use where someone does need to care.

Reply:
It would also be good to fix this instability upstream. Can't we argue that the off-diagonal elements are a strong zero like `false`, and make `NaN .* Diagonal(rand(3))` just work?

Reply:
They should be strong zeros yes.
that is the behavour suitesparse takes.
and julia uses that, and displays that behavour for SparseMatrixCSC.
I think that all structural zeros should be strong.
I would say it is a bug if it doesn't.

Reply:
I feel like all structural zeros should be strong yes.
I was sure I had seem julia displaying that behavour on SparseCSC matrixes, but I can't reproduce it right now.